doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+      vice versa.
+   6. Repeat steps 2.1 - 2.6 with using the both the deleted old and new app
+      passwords in step 2.2. Authentication using the deleted old app passwords
+      must fail, and authentication using the new app passwords must succeed.
+   7. Chceck that the user now only has the primary password.
+3. Test interactions with SSSD using the new functionality.
+   * Repeat all procedures from the point 2, but when testing logins to apps,
+     perform them on a client using SSSD configured against the testing FreeIPA
+     server.
+
+## Possible future enhancements
+
+### Support for use of the aforementioned functionality with Kerberos
+
+Support for use of the aforementioned functionality with Kerberos is necessary
+for HBAC to be applicable to users logged in using app passwords.
+
+Subaccounts will be introduced with principals in the form 
+```_user_/_appname_@_REALM_```
+(to be consistent with other things in FreeIPA) - a username will remain the
+same for subaccounts, only the principal will contain the app name. The code of
+FreeIPA must be modified to take into account the newly changed assumptions.
+Until now, these were:
+* two-component principals with a host as the first component are always host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are always service principals
+  (`_service_/_fqdn_`)
+* single-component principals are always users (user)
+Now, they will be:
+* two-component principals with an app name as the first component are users
+  logged in with an app password (`_user_/_appname_`)
+* two-component principals with a host as the first component are host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are service principals (`_service_/_fqdn_`)
+* single-component principals are users logged in with their primary password
+  (`user`)
+
+The user must not be able to create a principal looking like a host or a 
+service, or even set a password for another user. These problems are already
+addressed. The latter problem will be avoided by setting the username part of
+the principal automatically to the user's username. The former problem is
+already addressed - it is prevented by disallowing dots in the app name, which
+will be set by the user. The app name also must not be 'admin', as
+`_user_/_admin_@_REALM_` are principals of administrators. These constraints are
+enforced at the LDAP ACI level, not in the IPA Python frontend.
+
+The ipa-kdb daemon probably will not have to be extended to support this
+functionality, but it is possible.
+
+### HBAC for app passwords
+
+Two-component principals for subaccounts, as described above, will be used for
+assigning HBAC rules to users logged in using app passwords. These HBAC rules
+will be assigned by administrators just as the other ones.
+
+Hostgroups will be used to allow applying the same HBAC rule to different
+servers for the same app.
+
+HBAC itself will not need to be extended.
```

Comment:
I think libhbac would need to know how to connect an appname entry with its POSIX entry.

HBAC rules would need to be extended to add appname. If appname management is limited then I could see admins wanting to allow all appname=foo to be allowed to do something, as a form of grouping. Arguably that grouping could be done another way but I'd expect an RFE.

Reply:
HBAC for app passwords belongs to the optional 'advanced part' of my thesis assignment (as I wrote [above](https://github.com/freeipa/freeipa/pull/4061#discussion_r403455395) - you will need to unroll 4 hidden conversations).
We discussed it multiple times above, plus the most crucial decision to use HBAC for this part was made on a personal meeting between my supervisor, @stlaz , and @simo5 in January. But after it became clear that:
* it will require Kerberos support for app passwords, which is otherwise not needed,
* it will introduce a necessity to explicitly bar access to account management (logging directly into FreeIPA) for app passwords, which is now done implicitly by no Kerberos support for app passwords,
* I would definitely not be able to submit my thesis on time, if I would implement this as well,
I relegated this 'advanced part' to only a future RFE, which I will describe in the text of my thesis (based on our elaborations), but won't implement now.

Reply:
By libhbac, you mean `ipaserver/plugins/hbac*`? If HBAC itself would need to be extended, then it just adds to the reasons why to leave this for a future RFE. However, it seems that you wrote this when you assumed that I need to make a separate plugin for apps, what is no longer the case. Grouping in HBAC rules is planned, and I summarized ways to go and their advantages and disadvantages in the table in [this post](https://github.com/freeipa/freeipa/pull/4061#issuecomment-613377502).